pkg/controller/cronjob/utils.go

```diff
@@ -57,38 +69,78 @@ func deleteFromActiveList(cj *batchv1.CronJob, uid types.UID) {
 	cj.Status.Active = newActive
 }
 
-// getNextScheduleTime gets the time of next schedule after last scheduled and before now
-//  it returns nil if no unmet schedule times.
-// If there are too many (>100) unstarted times, it will raise a warning and but still return
-// the list of missed times.
-func getNextScheduleTime(cj batchv1.CronJob, now time.Time, schedule cron.Schedule, recorder record.EventRecorder) (*time.Time, error) {
-	var (
-		earliestTime time.Time
-	)
+// mostRecentScheduleTime returns:
+// - the last schedule time or CronJob's creation time,
+// - the most recent time a Job should be created or nil, if that's before now,
+// - number of missed schedules
+// - error in an edge case where the schedule specification is gramatically correct,
+//   but logically doesn't make sense (31st day for months with only 30 days, for example).
+func mostRecentScheduleTime(cj *batchv1.CronJob, now time.Time, schedule cron.Schedule, includeStartingDeadlineSeconds bool) (time.Time, *time.Time, int64, error) {
+	earliestTime := cj.ObjectMeta.CreationTimestamp.Time
 	if cj.Status.LastScheduleTime != nil {
 		earliestTime = cj.Status.LastScheduleTime.Time
-	} else {
-		// If none found, then this is either a recently created cronJob,
-		// or the active/completed info was somehow lost (contract for status
-		// in kubernetes says it may need to be recreated), or that we have
-		// started a job, but have not noticed it yet (distributed systems can
-		// have arbitrary delays).  In any case, use the creation time of the
-		// CronJob as last known start time.
-		earliestTime = cj.ObjectMeta.CreationTimestamp.Time
-	}
-	if cj.Spec.StartingDeadlineSeconds != nil {
-		// Controller is not going to schedule anything below this point
+	}
+	if includeStartingDeadlineSeconds && cj.Spec.StartingDeadlineSeconds != nil {
```

Comment:
Now that `startingDeadlineSeconds` isn't included in these calculations by default, would you mind un-commenting the test in https://github.com/kubernetes/kubernetes/blob/7f920da4420b72a30579975e98d668e0910d2df5/pkg/controller/cronjob/cronjob_controllerv2_test.go#L1052?

Reply:
For now I've kept the previous logic untouched so this case will still fail. It's something I'd leave for a followup to fix.